### PR TITLE
cmd/govim: check quickfix list origin before update

### PR DESCRIPTION
**Pull request by tbruyelle**
QuickfixDiagnostics routine updates the quickfix list each time it
receives new diagnostics from gopls. This can be problematic if the
quickfix list has been populated by an other plugin, because the user
could be using this plugin.

The fix uses the quickfix title to determine if the quickfix is "owned"
by govim or by an other plugin. If the quicfix title is empty (which
means the quickfix has never been used before) or equal to "govim", then
govim considers it can change it, else it considers it cannot.

This condition is ignored when :GOVIMQuickfixDiagnostics is invoked by
the user, making possible for the user to display again the diagnostics
when he wants to.

### CODE REVIEW

File: cmd/govim/quickfix.go
Comment (tbruyelle) [inline cmd/govim/quickfix.go:49]: On thing I would like to improve is to merge this condition and the one from `v.quickfixIsDiagnostics`.

The later is set to false when Govim populates the quickfix with references, but we could use quickfix title also in that case.

I think if I turn `v.quickfixIsDiagnostics` into a function that returns true if the quickfix title is empty or equal to "govim" (maybe rename that to "govim-diagnostics", because we could add an other title for Govim references), then it would clean a little bit the code :

- only one condition here instead of 2
- no more manual `v.quickfixIsDiagnostics = true|false` since it relies on quickfix title

File: cmd/govim/testdata/scenario_default/quickfix_from_other.txt
Comment (myitcv) [inline cmd/govim/testdata/scenario_default/quickfix_from_other.txt:4]: This is a great test.

One thing I think we need to do before calling this is verify that the quickfix list gets populated with the error diagnostics from `gopls`. That's not essential to this test, but instead a more realistic reflection of real life. So that way you end up with a call to `vimexprwait` before this call, then as you have it another one afterwards.

File: cmd/govim/testdata/scenario_default/quickfix_from_other.txt
Comment (myitcv) [inline cmd/govim/testdata/scenario_default/quickfix_from_other.txt:6]: If you put a blank line before this comment, [`testscript`](https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript) will treat it as a section. In the verbose output from `testscript`, sections that complete "successfully" are collapse under the "heading" of the comment that precedes the contiguous lines of that section.

File: cmd/govim/testdata/scenario_default/quickfix_from_other.txt
Comment (myitcv) [inline cmd/govim/testdata/scenario_default/quickfix_from_other.txt:10]: Ditto.

File: cmd/govim/quickfix.go
Comment (myitcv) [reply cmd/govim/quickfix.go:49]: Yes I'd be happy to revisit this. Either in this PR or a later one. As you observed, it's rather clumsy now. Vim is essentially the source of truth for what is in the quickfix window at any given time, and your change helps move us closer to that arrangement.

**Review by myitcv — COMMENTED** [review]
Generally LGTM with a couple of suggestions

cc @leitzler for thoughts too.

File: cmd/govim/testdata/scenario_default/quickfix_from_other.txt
Comment (tbruyelle) [reply cmd/govim/testdata/scenario_default/quickfix_from_other.txt:5]: @myitcv OK, I added an assertion for the first error 35c475d

File: cmd/govim/testdata/scenario_default/quickfix_from_other.txt
Comment (tbruyelle) [reply cmd/govim/testdata/scenario_default/quickfix_from_other.txt:8]: done 35c475d

File: cmd/govim/testdata/scenario_default/quickfix_from_other.txt
Comment (tbruyelle) [reply cmd/govim/testdata/scenario_default/quickfix_from_other.txt:13]: done 35c475d

**Review by myitcv — APPROVED** [review]
LGTM - will leave it up to you @tbruyelle if you want to make the change you discussed in this PR or a follow up.

Thank you for your patience with this change!

**Comment by tbruyelle** [comment]
@myitcv thx for the review. I'll try to see if I can implement the change we discuss simply, if yes I'll add a commit, if not I'll probably go with an other MR.

File: cmd/govim/quickfix.go
Comment (myitcv) [inline cmd/govim/quickfix.go:41]: Please change this from a `var` block to short variable declarations.

**Review by myitcv — COMMENTED** [review]
Thanks @tbruyelle - broadly speaking this looks fine, with the one requested changed for short variable declarations please. 

However, I'd like @leitzler to opine too, because this potentially influences our strategy with regards to a Vim grep workaround.

File: cmd/govim/quickfix.go
Comment (tbruyelle) [reply cmd/govim/quickfix.go:41]: @myitcv sure no problem, by curiosity why do you prefer this kind of declaration ? On my side I like to pack variable declarations like that, because that formats the code and I found it's more clear and readable.

File: cmd/govim/quickfix.go
Comment (leitzler) [inline cmd/govim/quickfix.go:47]: Really nits, but I personally think:
```go
if !force && (!diagsHasChanged || !quickfixIsDiagnostics) {
```
would be slightly more readable and consistent with other force checks in the code.

**Review by leitzler — APPROVED** [review]
First of all, nice feature!

When it comes to vimgrep this slightly alters the behaviour (apart from the obvious, that the existing matches doesn't disappear when diagnostics arrive). The behaviour that do change is if the quickfix list is "owned" by someone but empty. 

e.g.
1) There are no govim diagnostics
2) User calls vimgrep but finds no matches, it returns with a `E480: No match: foobar` message
3) The user introduce a new error, but the quickfix remains empty since the title changed, even if it wasn't populated

In my workflow I use `:cw` to open the quickfix list, and it might be confusing if it "doesn't work" as it isn't clear that the title changed at step 2. I think it make sense if `govim` actually did hi-jack the quickfix if the list is empty.

Apart from that and a tiny nit it LGTM.

File: cmd/govim/quickfix.go
Comment (tbruyelle) [reply cmd/govim/quickfix.go]: @leitzler ofc ! I need to revise my boolean algebra courses :)

d3c8ef1

**Comment by tbruyelle** [comment]
> In my workflow I use `:cw` to open the quickfix list, and it might be confusing if it "doesn't work" as it isn't clear that the title changed at step 2. I think it make sense if `govim` actually did hi-jack the quickfix if the list is empty.

@leitzler What if the new `isQuickfixDiagnostics()` function returns true if the quickfix list is empty ?

Maybe we should rename that function seems it's not really `isQuickfixDiagnostics` but more likely `canQuickfixDiagnostics`... WDYT?

File: cmd/govim/quickfix.go
Comment (myitcv) [reply cmd/govim/quickfix.go:41]: I've honestly not seen much (any?) use of parenthesised `var` declaration lists "outside" of top level file declarations. At the file level they do have things to stand out more, and when combined with `import` and `const` declarations, visually make sense. But within a function, they're a bit clumsy and unnecessarily verbose to my mind: which is precisely what short variable declarations exist to help with (amongst other situations like `for` loops).

**Comment by tbruyelle** [comment]
@leitzler @myitcv While I was trying some other simplifications in `updateQuickfixWithDiagnostics`,  I've spotted a bug in this change :

If I remove those lines : 
```diff
 func (v *vimstate) updateQuickfixWithDiagnostics(force bool) error {
-       if !force && (v.config.QuickfixAutoDiagnostics == nil || !*v.config.QuickfixAutoDiagnostics) {
-               return nil
-       }
```

The test ` TestScripts/scenario_default/quickfix_config` is still green, while it should be red, because `quickfixAutoDiagnostics` is ignored.

To make it red as it should be, I have to update the code in `vimstate.go`, which handles the config change of this particular field. Indeed, the code was updating the quickfix list with empty diagnostics when `autoDiagnostics` was set to false, but without setting the quickfix title. If it set the title as expected, the test goes red.

So the fix introduces a new function `setQuickfixDiagnostics` that populates the quickfix with diagnostics passed in parameters and sets the expected title.

See c977523

File: cmd/govim/quickfix.go
Comment (tbruyelle) [reply cmd/govim/quickfix.go]: Ok, as you wish b65554c

**Comment by myitcv** [comment]
@tbruyelle thanks for making the various changes. I'll hand over to @leitzler at this point for you both to round out the discussion above.

**Comment by leitzler** [comment]
> > In my workflow I use `:cw` to open the quickfix list, and it might be confusing if it "doesn't work" as it isn't clear that the title changed at step 2. I think it make sense if `govim` actually did hi-jack the quickfix if the list is empty.
> 
> @leitzler What if the new `quickfixIsDiagnostics()` function returns true if the quickfix list is empty ?

Yes, sounds good.

> Maybe we should rename that function seems it's not really `quickfixIsDiagnostics` but more likely `quickfixCanDiagnostics`... WDYT?

I agree, that would describe the function better.

**Comment by tbruyelle** [comment]
> > > In my workflow I use `:cw` to open the quickfix list, and it might be confusing if it "doesn't work" as it isn't clear that the title changed at step 2. I think it make sense if `govim` actually did hi-jack the quickfix if the list is empty.
> > 
> > 
> > @leitzler What if the new `quickfixIsDiagnostics()` function returns true if the quickfix list is empty ?
> 
> Yes, sounds good.
> 
> > Maybe we should rename that function seems it's not really `quickfixIsDiagnostics` but more likely `quickfixCanDiagnostics`... WDYT?
> 
> I agree, that would describe the function better.

@leitzler I have updated the function accordingly, and added a test. See 4edf3be

Now if the quickfix is empty, it can be updated by diagnostics, whatever its title.

**Comment by leitzler** [comment]
Looks good to me!

Please squash the commits as well before merge (as per https://github.com/govim/govim/wiki/Contributors#pull-request-contributions) .

Thank you very much for all the updates (and the patience)!

**Comment by tbruyelle** [comment]
> Looks good to me!
> 
> Please squash the commits as well before merge (as per https://github.com/govim/govim/wiki/Contributors#pull-request-contributions) .

done 2ed54c7 

> Thank you very much for all the updates (and the patience)!

No problem, it was a pleasure to contribute to this great project :)

**Pull request merged**
